x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchSingleNodeTests.java

```diff
@@ -0,0 +1,134 @@
+        for (ShardSearchFailure shardSearchFailure : phaseExecutionException.shardFailures()) {
+            assertEquals("boom", shardSearchFailure.getCause().getMessage());
+        }
+    }
+
+    public void testFetchFailuresOnlySomeShards() throws Exception {
+        for (int i = 0; i < 5; i++) {
+            IndexResponse indexResponse = client().index(new IndexRequest("boom" + i).id("boom" + i).source("text", "value")).get();
+            assertEquals(RestStatus.CREATED, indexResponse.status());
+        }
+        for (int i = 0; i < 5; i++) {
+            IndexResponse indexResponse = client().index(new IndexRequest("index" + i).id("index" + i).source("text", "value")).get();
+            assertEquals(RestStatus.CREATED, indexResponse.status());
+        }
+        client().admin().indices().refresh(new RefreshRequest()).get();
+
+        TermsAggregationBuilder agg = new TermsAggregationBuilder("text").field("text.keyword");
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder().aggregation(agg);
+        SubmitAsyncSearchRequest submitAsyncSearchRequest = new SubmitAsyncSearchRequest(sourceBuilder);
+        submitAsyncSearchRequest.setWaitForCompletionTimeout(TimeValue.timeValueSeconds(10));
+        AsyncSearchResponse asyncSearchResponse = client().execute(SubmitAsyncSearchAction.INSTANCE, submitAsyncSearchRequest).actionGet();
+
+        assertFalse(asyncSearchResponse.isRunning());
+        assertFalse(asyncSearchResponse.isPartial());
+        assertNull(asyncSearchResponse.getFailure());
+        SearchResponse searchResponse = asyncSearchResponse.getSearchResponse();
+        assertEquals(10, searchResponse.getTotalShards());
+        assertEquals(5, searchResponse.getSuccessfulShards());
+        assertEquals(5, searchResponse.getFailedShards());
+        assertEquals(10, searchResponse.getHits().getTotalHits().value);
+        assertEquals(5, searchResponse.getHits().getHits().length);
+        StringTerms terms = searchResponse.getAggregations().get("text");
+        assertEquals(1, terms.getBuckets().size());
+        assertEquals(10, terms.getBucketByKey("value").getDocCount());
+        assertEquals(5, searchResponse.getShardFailures().length);
+        for (ShardSearchFailure shardFailure : searchResponse.getShardFailures()) {
+            assertEquals("boom", shardFailure.getCause().getMessage());
+        }
+    }
+
+    public static final class SubFetchPhasePlugin extends Plugin implements SearchPlugin {
+        @Override
+        public List<FetchSubPhase> getFetchSubPhases(FetchPhaseConstructionContext context) {
+            return Collections.singletonList(searchContext -> new FetchSubPhaseProcessor() {
+                @Override
+                public void setNextReader(LeafReaderContext readerContext) {}
+
+                @Override
+                public void process(FetchSubPhase.HitContext hitContext) {
+                    if (hitContext.hit().getId().startsWith("boom")) {
```

Comment:
nit: you could derive it from the index name instead ?

Reply:
that is what I thought too but it turns out the index is null, no idea why.

Reply:
do you mean `hitContext.hit().getIndex()` is null ?

Reply:
yes pretty much everything is null in the hit besides the id